azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/AzureComputeServiceAdapter.java

```diff
@@ -0,0 +1,345 @@
+
+      logger.debug("Deployment created with name: %s", name);
+
+      final Set<VMDeployment> deployments = Sets.newHashSet();
+      if (!retry(new Predicate<String>() {
+         @Override
+         public boolean apply(final String name) {
+            runningNumber++;
+
+
+            ResourceGroupApi resourceGroupApi = api.getResourceGroupApi();
+            ResourceGroup resourceGroup = resourceGroupApi.get(getGroupId());
+            String resourceGroupName;
+
+            if (resourceGroup == null){
+               final Map<String, String> tags = ImmutableMap.of("description", "jClouds managed VMs");
+               resourceGroupName = resourceGroupApi.create(getGroupId(), location, tags).name();
+            } else {
+               resourceGroupName = resourceGroup.name();
+            }
+
+            Deployment deployment = api.getDeploymentApi(resourceGroupName).create(name, deploymentTemplate);
+
+            if (deployment != null) {
+               VMDeployment vmDeployment = new VMDeployment();
+               vmDeployment.deployment = deployment;
+               deployments.add(vmDeployment);
+            } else {
+               logger.debug("Failed to create deployment!");
+            }
+            return !deployments.isEmpty();
+         }
+      }, azureComputeConstants.operationTimeout(), 1, SECONDS).apply(name)) {
+         final String illegalStateExceptionMessage = format("Deployment %s was not created within %sms so it will be destroyed.",
+                 name, azureComputeConstants.operationTimeout());
+         logger.warn(illegalStateExceptionMessage);
+
+         throw new IllegalStateException(illegalStateExceptionMessage);
+      }
+
+      final VMDeployment deployment = deployments.iterator().next();
+
+
+      return new NodeAndInitialCredentials<VMDeployment>(deployment, name,
+              LoginCredentials.builder().user(loginUser).identity(loginUser).password(loginPassword).authenticateSudo(true).build());
+   }
+
+   @Override
+   public Iterable<VMSize> listHardwareProfiles() {
+      return api.getVMSizeApi(getLocation()).list();
+   }
+
+   private void getImagesFromPublisher(String publisherName, List<ImageReference> osImagesRef) {
+      OSImageApi osImageApi = api.getOSImageApi(getLocation());
+      Iterable<Offer> offerList = osImageApi.listOffers(publisherName);
+      for (Offer offer : offerList) {
+         Iterable<SKU> skuList = osImageApi.listSKUs(publisherName, offer.name());
+         for (SKU sku : skuList) {
+            osImagesRef.add(ImageReference.create(publisherName, offer.name(), sku.name(), null));
+         }
+      }
+   }
+
+   @Override
+   public Iterable<ImageReference> listImages() {
+      final List<ImageReference> osImages = Lists.newArrayList();
+      getImagesFromPublisher("Microsoft.WindowsAzure.Compute", osImages);
+      getImagesFromPublisher("MicrosoftWindowsServer", osImages);
+      getImagesFromPublisher("Canonical", osImages);
```

Comment:
It should be better to provide a property with the comma separated values of publishers. The api metadata should set the default value to these three elements, but this way we allow users to customize the list of publishers they want to use, when creating the context.

The same contract for locations applies here. After being transformed by the transformation function, global images must not have the location set, and images that only exist in a particular location must inform it.